lib/stats/stats-query.c

```diff
@@ -117,62 +54,74 @@ _is_single_match(const gchar *key_str)
   return !strpbrk(key_str, "*?");
 }
 
-static GList *
-_query_counter_hash(const gchar *key_str)
+static void
+_process_counter_if_matching(StatsCluster *sc, gint type, StatsCounterItem *counter,
+                             gpointer user_data /*, gboolean* cancelled*/)
 {
-  GPatternSpec *pattern = g_pattern_spec_new(key_str);
-  GList *counters = NULL;
-  gpointer key, value;
-  GHashTableIter iter;
-  gboolean single_match;
-
-  _update_index();
-  single_match = _is_single_match(key_str);
-
-  g_mutex_lock(&stats_query_mutex);
-  g_hash_table_iter_init(&iter, counter_index);
-  while (g_hash_table_iter_next(&iter, &key, &value))
+  gpointer *args = (gpointer *) user_data;
+  int arg_ndx = 0;
+  GPatternSpec *pattern = (GPatternSpec *) args[arg_ndx++];
+  gboolean single_match = *(gboolean const *) args[arg_ndx++];
+  ProcessCounterCb process_func = (ProcessCounterCb) args[arg_ndx++];
+  gpointer process_func_user_data = args[arg_ndx++];
+  StatsFormatCb format_cb = (StatsFormatCb) args[arg_ndx++];
```

Comment:
This patch `stats: added and using now a common implementation of stat counter reset ` contains multiple refactors steps as well. Please consider extracting them next time, it helps a lot to prove equivalence during review.